core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -468,21 +468,20 @@ private[spark] class DAGScheduler(
 
   /** Find ancestor shuffle dependencies that are not registered in shuffleToMapStage yet */
   private def getMissingAncestorShuffleDependencies(
-      rdd: RDD[_]): ArrayStack[ShuffleDependency[_, _, _]] = {
-    val ancestors = new ArrayStack[ShuffleDependency[_, _, _]]
+      rdd: RDD[_]): ListBuffer[ShuffleDependency[_, _, _]] = {
+    val ancestors = new ListBuffer[ShuffleDependency[_, _, _]]
     val visited = new HashSet[RDD[_]]
     // We are manually maintaining a stack here to prevent StackOverflowError
     // caused by recursively visiting
-    val waitingForVisit = new ArrayStack[RDD[_]]
-    waitingForVisit.push(rdd)
+    val waitingForVisit = ListBuffer[RDD[_]](rdd)
     while (waitingForVisit.nonEmpty) {
-      val toVisit = waitingForVisit.pop()
+      val toVisit = waitingForVisit.remove(0)
       if (!visited(toVisit)) {
         visited += toVisit
         getShuffleDependencies(toVisit).foreach { shuffleDep =>
           if (!shuffleIdToMapStage.contains(shuffleDep.shuffleId)) {
-            ancestors.push(shuffleDep)
-            waitingForVisit.push(shuffleDep.rdd)
+            ancestors.+=:(shuffleDep)
```

Comment:
"push" an element on the front. We can also call `prepend` which is clearer but has some varargs overhead. This operator is weird and IntelliJ claims it can't be invoked like `foo +=: bar`, but I wonder if it's mistaken. I don't so much like this invocation.

Reply:
You can write:
```scala
shuffleDep +=: ancestors
```
Just in case, any operator ending in `:` is right-associative in Scala

Reply:
Yep got it now; I'm going with prepend.